maintenance/health/servicehealthcheck/healthchecker.go

```diff
@@ -0,0 +1,138 @@
+	HealthCheck() (bool, error)
+	CleanUp() error
+}
+
+// ConnectionState can be used for Health Checks
+// It offers a Mutex and the Date of the last check for caching the result
+type ConnectionState struct {
+	LastCheck time.Time
+	isHealthy bool
+	err       error
+	m         sync.Mutex
+}
+
+type handler struct{}
+
+// checks map with all health checks, key: Name of the check
+var checks = make(map[string]HealthCheck)
+var checksLock = sync.RWMutex{}
+
+// initErrors map with all errors that happened in the initialisation of the health checks
+var initErrors = make(map[string]error)
+var router *mux.Router
+
+func (cs *ConnectionState) setConnectionState(healthy bool, err error, mom time.Time) {
+	cs.m.Lock()
+	defer cs.m.Unlock()
+	cs.isHealthy = healthy
+	cs.err = err
+	cs.LastCheck = mom
+}
+
+func (cs *ConnectionState) SetErrorState(err error, mom time.Time) {
+	cs.setConnectionState(err == nil, err, mom)
+}
+
+func (cs *ConnectionState) SetHealthy(mom time.Time) {
+	cs.setConnectionState(true, nil, mom)
+}
+
+func (cs *ConnectionState) GetState() (bool, error) {
+	return cs.isHealthy, cs.err
+}
+
+func NewConnectionState() *ConnectionState {
+	return &ConnectionState{m: sync.Mutex{}}
+}
+
+func (h *handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	route := r.URL.Path
+	splitRoute := strings.Split(route, "/health/")
+
+	//Check the route first
+	if len(splitRoute) < 2 || splitRoute[0] != "" {
+		h.writeError(w, errors.New("Route not Valid\n"), http.StatusBadRequest, route)
+		return
+	}
+
+	checksLock.RLock()
+	defer checksLock.RUnlock()
+	name := splitRoute[1]
+
+	if checks[name] == nil {
+		h.writeError(w, errors.New("Health Check not registered\n"), http.StatusNotFound, name)
+		return
+	}
+	hc := checks[name]
+	// If it was not possible to initialise this health check, then show the initialisation error
+	if err := initErrors[name]; err != nil {
+		h.writeError(w, err, http.StatusServiceUnavailable, name)
+		return
+	}
+
+	// this is the actual health check
+	if healthy, err := hc.HealthCheck(); healthy {
+		// to increase performance of the request set
+		// content type and write status code explicitly
+		w.Header().Set("Content-Type", "text/plain")
+		w.WriteHeader(http.StatusOK)
+		w.Write([]byte("OK\n")) // nolint: gosec,errcheck
+	} else {
+		h.writeError(w, err, http.StatusServiceUnavailable, name)
+	}
+}
+
+func (h *handler) writeError(w http.ResponseWriter, err error, errorCode int, name string) {
+	w.Header().Set("Content-Type", "text/plain")
+	w.WriteHeader(errorCode)
+	w.Write([]byte("Error \n")) // nolint: gosec,errcheck
+	log.Warnf("helthcheck %v was not healthy %v", name, err)
+}
+
+// RegisterHealthCheck register a healthCheck that need to be routed
+// names must be uniq
+func RegisterHealthCheck(hc HealthCheck, name string) {
+	if checks[name] != nil {
+		log.Debugf("Health checks can only be added once, tried to add another health check with name %v", name)
+		return
+	}
+	if router == nil {
```

Comment:
don't do nil checking, why would it be nil, isn't this a panic?

Reply:
O don't think so because then health checks could make an application unusable. Changed the logging Level to Error.